streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java

```diff
@@ -184,6 +203,52 @@ public void process(final Record<KIn, VIn> record) {
                     droppedRecordsSensor.record();
                 }
             }
+
+            if (emitFinal) {
+                if (!(windows instanceof TimeWindows)) {
+                    throw new IllegalStateException("Emit final is only supported for TimeWindows and SlidingWindows for TimeWindowedKStream");
+                }
+
+                // Close time does not progress
+                if (lastEmitCloseTime >= closeTime) {
+                    return;
+                }
+
+                // Emit final deosn't make sense for some window like unlimitedWindow
+                // This should be checked in upstream before pass emit final in or checked here
+                final long closeWindowStart = closeTime - windows.size();
+                if (closeWindowStart <= 0) {
+                    return;
+                }
+
+                final long lastEmitWindowStart = lastEmitCloseTime == ConsumerRecord.NO_TIMESTAMP ?
+                    0 : lastEmitCloseTime - windows.size();
+
+                final Map<Long, W> matchedCloseWindows = windows.windowsFor(closeWindowStart);
+                final Map<Long, W> matchedEmitWindows = windows.windowsFor(lastEmitWindowStart);
+
+                // Don't fetch store if there's new close time doesn't progress enough to cover next
+                // window
+                if (matchedCloseWindows.equals(matchedEmitWindows)) {
+                    log.debug("no new windows to emit. LastEmitCloseTime={}, newCloseTime={}", lastEmitCloseTime, closeTime);
+                    return;
+                }
+
+                KeyValueIterator<Windowed<KIn>, ValueAndTimestamp<VAgg>> windowToEmit =  windowStore
+                    .fetchAll(lastEmitWindowStart + 1, closeWindowStart);
+
+                while (windowToEmit.hasNext()) {
+                    KeyValue<Windowed<KIn>, ValueAndTimestamp<VAgg>> kv = windowToEmit.next();
+                    tupleForwarder.maybeForward(
+                        record.withKey(kv.key)
+                            .withValue(new Change<>(kv.value.value(), null))
```

Comment:
This is not a suggestion: we are sending `null` with the guarantee that we should have never forward for this key before. I think a good test case coverage would be to have a windowed aggregation emit final, followed by a join. The join results would need both the old/new values to be able to correct, and if emit final we should only emit once, with old value setting to null.